File: python/graph.py
from hash_table import HashTable
from vertex import Vertex
from edge import Edge

class Graph(object):
    def __init__(self):
        self.vertices = HashTable(20)

    # creates a vertex from a location and adds to vertex hash table: O(n)
    def add_vertex(self, location):
        self.vertices.insert(location.identifier, Vertex(location))

    # creates a bi-directional weighted edge between two vertexes in the graph: O(n)
    def add_weighted_edge(self, origin, destination, weight):
        self.vertices.find(origin.identifier).add_edge(Edge(destination, weight))
        self.vertices.find(destination.identifier).add_edge(Edge(origin, weight))

    # finds the vertex matching the location: O(n)
    def find_vertex(self, location):
        return self.vertices.find(location.identifier)

    # finds the distance between two vertexes: O(n)
    def find_distance_between(self, origin, target):
        return self.vertices.find(origin.identifier).distance_to(target)

    # finds the distance between a location and where the package needs to be delivered
    # similar to the method above but it used to sort the priority lists
    def distance_to_deliver(self, location):
        def distance_to(package):
            return self.vertices.find(location.identifier).distance_to(package.destination)

        return distance_to

    # creates a closure/lambda that is used to find the distances between locations
    # this is used when finding the next closest location the truck should drive to
    def distance_from(self, origin):
        def distance_to(destination):
            return self.vertices.find(origin.identifier).distance_to(destination)

        return distance_to

File: ruby/graph.rb
class Graph
  def initialize
    @vertices = HashTable.new(20)
    @locations_hash = HashTable.new(20)
  end

  def add_vertex(location)
    vertices.insert(location.id, Vertex.new(location))
    locations_hash.insert(location.address.hash.abs, location)
  end

  def add_weighted_edge(origin, destination, weight)
    vertices.find(origin.id).add_edge(Edge.new(destination, weight))
    vertices.find(destination.id).add_edge(Edge.new(origin, weight))

    true
  end

  def find_location(address)
    locations_hash.find(address.hash.abs)
  end

  def find_distance(from, to)
    vertices.find(from.id).find_edge(to.id).weight
  end

  # def find_closest_neighbor_with_available_packages(location, truck)
  #   start_node = vertices.find(location.id)
  #   nodes_to_visit = start_node.edges.clone
  #   visited = HashTable.new(40)
  #   packages = []

  #   until nodes_to_visit.empty? || truck.full?  do
  #     node = nodes_to_visit.unshift
  #     
  #     
  #     closest = start_node.find_closest_neighbor_with_available_packages(truck)

  #   closest.nil? ? find_closest_neighbor_with_available_packages(start_node.closest_neighbor, truck) : closest.location
  # end

  private

  attr_accessor :vertices, :locations_hash
end
